### PR TITLE
[kueuectl] Support autocompletion on kueue CLI commands

### LINKED ISSUE #2295: [kueuectl] Support autocompletion on kueue CLI commands
**What would you like to be added**:
Support autocompletion on kueue CLI commands.

**Why is this needed**:

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Support autocompletion on kueue CLI commands.

#### Which issue(s) this PR fixes:

Fixes #2295 

#### Does this PR introduce a user-facing change?

```release-note
Support autocompletion on kueue CLI commands
```

### CODE REVIEW

**Review by trasc — COMMENTED** [review]
/lgtm

File: cmd/kueuectl/app/util/completion.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+		if err != nil {
+			return []string{}, cobra.ShellCompDirectiveError
+		}
+
+		namespace, _, err := clientGetter.ToRawKubeConfigLoader().Namespace()
+		if err != nil {
+			return []string{}, cobra.ShellCompDirectiveError
+		}
+
+		list, err := clientSet.KueueV1beta1().LocalQueues(namespace).List(cmd.Context(), metav1.ListOptions{})
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go:192]: Yeah, I thought about it. But the problem that when we will put name, cobra will filter it and not all localqueue names will suggest.
5 I think is very small limit. Maybe 100?

[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go:192]: @alculquicondor WDYT?

File: cmd/kueuectl/app/cmd.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ func NewDefaultKueuectlCmd() *cobra.Command {
 
 func NewKueuectlCmd(o KueuectlOptions) *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "kueue",
+		Use:   "kueuectl",
```
Comment (alculquicondor) [inline cmd/kueuectl/app/cmd.go:55]: why change this?

Will it work for autocompletion when you don't have an alias (`kubectl kueue`)?

File: cmd/kueuectl/app/util/fake.go
Comment (alculquicondor) [inline cmd/kueuectl/app/util/fake.go:1]: I think we can keep these objects in a `testing` package.

File: cmd/kueuectl/app/util/completion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply cmd/kueuectl/app/util/completion.go:192]: Actually, do these show up in the terminal? If not, maybe 100 is ok. Or does this depend on the terminal?

File: cmd/kueuectl/app/util/completion.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		}
+
+		list, err := clientSet.CoreV1().Namespaces().List(cmd.Context(), metav1.ListOptions{})
+		if err != nil {
+			return []string{}, cobra.ShellCompDirectiveError
+		}
+
+		validArgs := make([]string, len(list.Items))
+		for i, wl := range list.Items {
+			validArgs[i] = wl.Name
```
Comment (alculquicondor) [inline cmd/kueuectl/app/util/completion.go:43]: Don't you also need to add `if strings.HasPrefix(name, toComplete)`?

File: cmd/kueuectl/app/util/completion.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			for _, wl := range list.Items {
+				if ptr.Deref(wl.Spec.Active, true) == *status {
+					filteredItems = append(filteredItems, wl)
+				}
+			}
+			list.Items = filteredItems
+		}
+
+		validArgs := make([]string, len(list.Items))
+		for i, wl := range list.Items {
```
Comment (alculquicondor) [inline cmd/kueuectl/app/util/completion.go:130]: same question whether you need `if strings.HasPrefix(name, toComplete)`

And same for the rest of the list calls

File: cmd/kueuectl/app/util/completion.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+		for name := range config.AuthInfos {
+			if strings.HasPrefix(name, toComplete) {
+				validArgs = append(validArgs, name)
+			}
+		}
+		return validArgs, cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+func WorkloadNameCompletionFunc(clientGetter ClientGetter, status *bool) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```
Comment (alculquicondor) [inline cmd/kueuectl/app/util/completion.go:98]:
```suggestion
func WorkloadNameCompletionFunc(clientGetter ClientGetter, activeStatus *bool) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```

File: cmd/kueuectl/app/util/completion.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+		validArgs := make([]string, len(list.Items))
+		for i, wl := range list.Items {
+			validArgs[i] = wl.Name
+		}
+
+		return validArgs, cobra.ShellCompDirectiveNoFileComp
+	}
+}
+
+func ClusterQueueNameCompletionFunc(clientGetter ClientGetter, status *bool) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```
Comment (alculquicondor) [inline cmd/kueuectl/app/util/completion.go:138]:
```suggestion
func ClusterQueueNameCompletionFunc(clientGetter ClientGetter, stopPolicy *bool) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```

File: cmd/kueuectl/app/cmd.go (hunk 2 above)
Comment (mbobrovskyi) [reply cmd/kueuectl/app/cmd.go:55]: No, it's not working with `kubectl kueue`. 
Also it's not working with `kueuectl` without this changes

File: cmd/kueuectl/app/util/fake.go
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/fake.go:1]: In this case we have cycle imports issue due to we have `completion_test.go`. :(

File: cmd/kueuectl/app/util/completion.go
Comment (alculquicondor) [inline cmd/kueuectl/app/util/completion.go:1]: `app/util/completion/completion.go`

File: cmd/kueuectl/app/cmd.go (hunk 2 above)
Comment (alculquicondor) [reply cmd/kueuectl/app/cmd.go:55]: How to make it work for `kubectl kueue`? That will be the first way users will use this binary.

Then we can check how we can make it support both.

File: cmd/kueuectl/app/util/completion.go (hunk 3 above)
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go:43]: No, it's not necessary. Cobra filtering it.

File: cmd/kueuectl/app/util/completion.go (hunk 6 above)
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go:138]: Not sure that `stopPolicy`... Maybe `activeStatus` too?

File: cmd/kueuectl/app/util/completion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go:192]: It's depends of terminal size. Cobra will show all items that can fit. But if we will set limit to 5 we can show only this 5 items.

File: cmd/kueuectl/app/cmd.go (hunk 2 above)
Comment (mbobrovskyi) [reply cmd/kueuectl/app/cmd.go:56]: For `kubectl kueue` we need to add `kubectl_complete-<pluginName>` file to $PATH directory. I will add it to docs.

File: cmd/kueuectl/app/util/completion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/kueuectl/app/util/completion.go]: Set limitation to 100.

File: cmd/kueuectl/app/completion/completion.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	"sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/cmd/kueuectl/app/util"
+)
+
+const completionLimit = 100
+
+func NamespaceNameCompletionFunc(clientGetter util.ClientGetter) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```
Comment (alculquicondor) [inline cmd/kueuectl/app/completion/completion.go:32]: nit:
```suggestion
func NamespaceNameFunc(clientGetter util.ClientGetter) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
```

Given that `completion` is already the package name.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
/hold
for nit

File: cmd/kueuectl/app/completion/completion.go (hunk 7 above)
Comment (mbobrovskyi) [reply cmd/kueuectl/app/completion/completion.go]: Fixed

**Comment by alculquicondor** [comment]
Please check integration failure

**Comment by mbobrovskyi** [comment]
@alculquicondor Looks like flakiness problem [here](https://github.com/kubernetes-sigs/kueue/blob/main/test/integration/kueuectl/list_test.go#L207).


`[FAILED] Expected <string>: "...     1s" to equal               |  <string>: "...     0s"`

 Should I fix here or on another PR?

**Comment by alculquicondor** [comment]
oh ok, yes, fix in a separate PR

/lgtm

**Comment by mbobrovskyi** [comment]
Created issue #2353

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
CLI: Support autocompletion
```

**Pull request merged**